qiskit/pulse/builder.py

```diff
@@ -731,27 +736,74 @@ def append_instruction(self, instruction: instructions.Instruction):
         """
         self._context_stack[-1].append(instruction)
 
+    def append_reference(self, name: str, *extra_keys: str):
+        """Add external program as a :class:`~qiskit.pulse.instructions.Reference` instruction.
+
+        Args:
+            name: Name of subroutine.
+            extra_keys: Assistance keys to uniquely specify the subroutine.
+        """
+        inst = instructions.Reference(name, *extra_keys)
+        self.append_instruction(inst)
+
     @_compile_lazy_circuit_before
     def append_block(self, context_block: ScheduleBlock):
         """Add a :class:`ScheduleBlock` to the builder's context schedule.
 
         Args:
             context_block: ScheduleBlock to append to the current context block.
+
+        Raises:
+            PulseError: When non ScheduleBlock object is appended.
         """
+        if not isinstance(context_block, ScheduleBlock):
+            raise exceptions.PulseError(
+                f"'{context_block.__class__.__name__}' is not valid data format in the builder. "
+                "Only 'ScheduleBlock' can be appended to the builder context."
+            )
+
         # ignore empty context
         if len(context_block) > 0:
             self._context_stack[-1].append(context_block)
 
-    def append_reference(self, name: str, *extra_keys: str):
-        """Add external program as a :class:`~qiskit.pulse.instructions.Reference` instruction.
+    @singledispatchmethod
+    def inject_subroutine(
+        self,
+        subroutine: Union[Schedule, ScheduleBlock],
+    ):
+        """Append a :class:`ScheduleBlock` to the builder's context schedule.
+
+        This operationd doesn't create reference. Subrotuine is directly
+        injected into current context schedule.
 
         Args:
-            name: Name of subroutine.
-            extra_keys: Assistance keys to uniquely specify the subroutine.
+            subroutine: ScheduleBlock to append to the current context block.
+
+        Raises:
+            PulseError: When subroutine is not Schedule nor ScheduleBlock.
         """
-        inst = instructions.Reference(name, *extra_keys)
-        self.append_instruction(inst)
+        raise exceptions.PulseError(
+            f"Subroutine type {subroutine.__class__.__name__} is "
+            "not valid data format. Inject Schedule or ScheduleBlock."
+        )
 
+    @inject_subroutine.register
```

Comment:
Naoki: if you're finding that you need to do non-trivial validation / setup before the actual dispatched code, you might want to borrow one of the co-operative subclassing patterns from standard OO (where not everything is magical virtual dispatch), and have:
```python
def inject_subroutine(self, schedule):
    # Do common setup / validation
    out = self._inject_subroutine(schedule)
    # Do common post-processing on the result
    return out

@dispatch
def _inject_subroutine(self, schedule): raise NotImplementedError
@_inject_subroutine.register
def _inject_subroutine(self, schedule: Schedule): ...
@_inject_subroutine.register
def _inject_subroutine(self, schedule: ScheduleBlock): ...
```

Daniel: in this case, when there's only two valid types and the body content is small, then it doesn't really matter too much (and I personally wouldn't use it here).  When the body content gets much longer and you start needing to factor it out into separate functions, or when the import chain starts getting long, then it can lead to much clearer code than long and deep `if ... elif` chains.

(For some people, it's also an advantage that other components of the library / users can register methods that act on their own types later on, though that's not a concern here.)

Reply:
In this one particular case, I personally wouldn't use dispatch at all - I'd just start the function with
```python
if isinstance(schedule, Schedule):
   schedule = self._naive_typecast_schedule(schedule)
```
and then do what it did before anyway.

Reply:
Fair enough. I personally like the idea of generic function, because this pattern illustrates the method has type-dependent behavior, rather than simply accepting two different types in the same way. Probably this is over engineering fore here. Removed in https://github.com/Qiskit/qiskit-terra/pull/8949/commits/0ca8efbc78481b5edd7456ec02b026a96785981a

Reply:
Thanks. I see the point of such a construct now but here it looks like over engineering.